subsys/net/lib/sockets/sockets.c

```diff
@@ -1337,13 +1385,19 @@ static inline ssize_t zsock_recv_dgram(struct net_context *ctx,
 	if (src_addr && addrlen) {
 		if (IS_ENABLED(CONFIG_NET_OFFLOAD) &&
 		    net_if_is_ip_offloaded(net_context_get_iface(ctx))) {
-			/*
-			 * Packets from offloaded IP stack do not have IP
-			 * headers, so src address cannot be figured out at this
-			 * point. The best we can do is returning remote address
-			 * if that was set using connect() call.
-			 */
-			if (ctx->flags & NET_CONTEXT_REMOTE_ADDR_SET) {
+			if (!net_pkt_addr_is_unspecified(pkt)) {
+				int rv;
+
+				rv = sock_get_offload_pkt_src_addr(pkt,
+							    src_addr, *addrlen);
+
+				if (rv < 0) {
+					errno = -rv;
+					LOG_ERR("sock_get_offload_pkt_src_addr %d",
+										rv);
+					goto fail;
+				}
+			} else if (ctx->flags & NET_CONTEXT_REMOTE_ADDR_SET) {
 				memcpy(src_addr, &ctx->remote,
 				       MIN(*addrlen, sizeof(ctx->remote)));
 			} else {
```

Comment:
I was actually thinking that all of the details about how address is obtained (including the fallback method) could be encapsulated witihn `sock_get_offload_pkt_src_addr()`, but current appraoch seems fine as well. no strong push to change from my side.

Reply:
I have restrucutred this to make it more consistent with the else state